### PR TITLE
Improvements for handling str.to_code

### PR DESCRIPTION
**Pull request by ajreynol**
This PR changes when we apply injectivity (see C-Inj https://homepage.divms.uiowa.edu/~ajreynol/ijcar20a.pdf) for str.to_code.

In particular, it should only be necessary to apply injectivity for pairs of string equivalence classes that may have length 1 and that are disequal.  Previously we applied it to add pairs of equivalence classes, apart from pairs that both contain constants.

This allows us to send significantly fewer lemmas in some cases and has a positive effect on a small number of benchmarks of interest. 

It also changes so we don't do a reduction lemma (range constraint) for str.to_code applied to concrete strings, since we have a concrete integer equality in these cases.

### CODE REVIEW

File: src/theory/strings/code_point_solver.cpp [hunk 1]
```diff
@@ -47,85 +54,95 @@ void CodePointSolver::checkCodes()
...
-  // str.code applied to the code term for each equivalence class that has a
-  // code term but is not a constant
-  std::vector<Node> nconst_codes;
-  // str.code applied to the proxy variables for each equivalence classes that
-  // are constants of size one
-  std::vector<Node> const_codes;
+  // We construct a mapping from string equivalent classes to code point
+  // applications. This mapping contains entries:
+  // (1) r -> (str.to_code t) where r is the representative of t and
+  // term and this str.to_code term occurs in the equality engine.
```
Comment (yoni206) [inline src/theory/strings/code_point_solver.cpp:60]: who is `term`?

File: src/theory/strings/term_registry.cpp [hunk 2]
```diff
@@ -163,6 +163,7 @@ Node TermRegistry::lengthPositive(Node t)
   Node tlenEqZero = tlen.eqNode(zero);
   Node tEqEmp = t.eqNode(emp);
   Node caseEmpty = nm->mkNode(AND, tlenEqZero, tEqEmp);
+  // Node caseEmpty = tEqEmp;
```
Comment (yoni206) [inline src/theory/strings/term_registry.cpp:166]: delete?

File: src/theory/strings/core_solver.h [hunk 3]
```diff
@@ -522,6 +524,8 @@ class CoreSolver : protected EnvObj
    * on the ordering described in checkCycles.
    */
   std::vector<Node> d_strings_eqc;
+  /** The relevant disequalities */
```
Comment (yoni206) [inline src/theory/strings/core_solver.h:527]:
```suggestion
  /** The relevant disequalities for `to_code`*/
```
?

Comment (yoni206) [reply src/theory/strings/core_solver.h:535]: Thanks for the changes. Only this one is left. If it is correct, I think it would make things clearer.

Comment (ajreynol) [reply src/theory/strings/core_solver.h:535]: This is documented above the public method to access this (see `getRelevantDeq`).

**Review by yoni206 — APPROVED** [review]

**Pull request merged**